### ISSUE TITLE
gtk3reactor freezes for seconds at a time (in trial)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5512]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5512|
|Type|defect|
|Created|2012-03-07 02:12:48Z|
|Branch|https://github.com/twisted/twisted/tree/gtk3-process-5512|

Running the twisted tests under gtk3 takes about 3 minutes longer on my machine than gtk2, mostly because it freezes for seconds on end on certain tests. It eventually gets unstuck, but it does seem less than optimal.

<details><summary>Searchable metadata</summary>

```
trac-id__5512 5512
type__defect defect
reporter__itamar itamar
priority__normal normal
milestone__Twisted_12_1 Twisted-12.1
branch__branches_gtk3_process_5512 branches/gtk3-process-5512
branch_author__itamar itamar
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1331086368000000 1331086368000000
changetime__1336830986000000 1336830986000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

This happens for gtk3, but not gi, in at least one test involving processes: I suspect SIGCHLD issues.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

This fixes the problem, and makes me strongly suspect that _GlibSignalMixin is broken and useless and should just be removed completely (although see [#4286](https://github.com/twisted/twisted/issues/4286) which implies there's some reason it exists):

```
Index: twisted/internet/_glibbase.py
===================================================================
--- twisted/internet/_glibbase.py	(revision 33656)
+++ twisted/internet/_glibbase.py	(working copy)
@@ -59,7 +59,7 @@
             # back to our extension module.  See #4286.
             from twisted.internet.process import (
                 reapAllProcesses as _reapAllProcesses)
-            base._SignalReactorMixin._handleSignals(self)
+            posixbase.PosixReactorBase._handleSignals(self)
             signal.signal(signal.SIGCHLD,
                           lambda *a: self.callFromThread(_reapAllProcesses))
             if getattr(signal, "siginterrupt", None) is not None:
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

(In [33868]) Branching to 'gtk3-process-5512'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst removed owner|
|-|-|

I've fixed this in the branch, also fixing an issue where "trial -r gtk2 twisted.internet.test.test_process" would wedge. I did this by changing to a more aggressive (and consistent) method of installing the SIGCHLD handler.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/gtk3-process-5512

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @itamarst|
|-|-|

Faster indeed! Looks great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set status to `closed`|
|-|-|

(In [34371]) Merge gtk3-process-5512: Notice SIGCHLD in gtk3reactor.

Fixes: [#5512](https://github.com/twisted/twisted/issues/5512)
Author: itamar
Review: therve


The previous method for signal handler installation in glib/gtk reactors wasn't sufficient in all cases, so switched to a simpler, less error-prone method.
